### PR TITLE
#1526 edit material endpoint

### PR DESCRIPTION
**Pull request by callmejim1226**
## Changes

Added material edit endpoint in 
- project routes
- project controller
- project service
Added tests for edit material endpoint

## Test Cases

- updateMaterial (6)
       ✓ Update material fails if the given material does not exists
       ✓ Update material fails if the given material deleted
       ✓ Update material fails if the project of the material does not exists
       ✓ Update material fails if the project of the material deleted
       ✓ Update material fails if the submitter is not a leader or part of project team
       ✓ Update material successfully works

## Screenshots

**Create new material for testing purposes:**
#1526-get-new-material

**The payload for the update material endpoint:**
image

**Update the material successfully:**
#1526-update-material

## To Do

_Any remaining things that need to get done_

- [x] Fix type conflict error on `assembly` and `quantityUnit` props in `Prisma.material.update`
- [x] Add tests for the edit material endpoint
- [x] Add screenshots of the successful run on endpoint from Postman

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: src/backend/src/routes/projects.routes.ts [hunk 1]
```diff
@@ -81,5 +81,25 @@ projectRouter.post(
...
+  intMinZero(body('assembly.assemblyId').optional()),
+  nonEmptyString(body('assembly.name').optional()),
+  nonEmptyString(body('assembly.pdmFileName').optional()),
+  isMaterialStatus(body('status')),
+  nonEmptyString(body('manufacturerName')),
+  nonEmptyString(body('manufacturerPartNumber')),
+  nonEmptyString(body('pdmFileName').optional()),
+  intMinZero(body('quantity')),
+  nonEmptyString(body('quantityUnit.name').optional()),
+  nonEmptyString(body('unitName')),
```
Comment (RChandler234) [inline src/backend/src/routes/projects.routes.ts:96]: should be optional

File: src/backend/src/services/projects.services.ts [hunk 2]
```diff
@@ -811,4 +811,87 @@ export default class ProjectsService {
...
+    status: Material_Status,
+    manufacturerName: string,
+    manufacturerPartNumber: string,
+    quantity: number,
+    unitName: string,
+    price: number,
+    subtotal: number,
+    linkUrl: string,
+    notes: string,
+    assembly?: Assembly,
```
Comment (RChandler234) [inline src/backend/src/services/projects.services.ts:844]: should be assemblyId - string

File: src/backend/src/services/projects.services.ts [hunk 3]
```diff
@@ -811,4 +811,87 @@ export default class ProjectsService {
...
+    pdmFileName?: string,
+    quantityUnit?: Unit
```
Comment (RChandler234) [inline src/backend/src/services/projects.services.ts:846]: this isn't an input, we get the unit, from the unitName which acts as the id for a unit

File: src/backend/src/services/projects.services.ts [hunk 4]
```diff
@@ -811,4 +811,87 @@ export default class ProjectsService {
...
+   */
+  static async editMaterial(
+    submitter: User,
+    wbsNumber: WbsNumber,
+    name: string,
+    status: Material_Status,
+    manufacturerName: string,
+    manufacturerPartNumber: string,
+    quantity: number,
+    unitName: string,
```
Comment (RChandler234) [inline src/backend/src/services/projects.services.ts:839]: should optional, this is what you use to connect the quantity unit, it's a bit confusing name

File: src/backend/src/services/projects.services.ts [hunk 5]
```diff
@@ -811,4 +811,87 @@ export default class ProjectsService {
...
+        wbsElementId: project.wbsElementId,
+        // assembly: {
+        //   connect:
+        //     {
+        //       assembly
+        //     } || undefined
+        // },
+        pdmFileName: pdmFileName || undefined
+        // quantityUnit: {
+        //   connect:
```
Comment (RChandler234) [inline src/backend/src/services/projects.services.ts:887]: I believe you can set assemblyId and unitName directly and it should work the same, while avoiding this issue you're having

**Review by RChandler234 — CHANGES_REQUESTED** [review]
look at the create material endpoint to make sure you're doing all the same checks it does there

File: src/backend/src/routes/projects.routes.ts [hunk 6]
```diff
@@ -81,5 +81,25 @@ projectRouter.post(
   validateInputs,
   ProjectsController.createMaterial
 );
+projectRouter.post(
+  '/bom/material/:wbsNum/edit',
```
Comment (RChandler234) [inline src/backend/src/routes/projects.routes.ts:85]: apologies, this should be :materialId, and then you should use that to fetch the material/ check if it exists in the service method. You should be able to get the wbsNum from that material that you fetch, it shouldn't be changing after a material has been created

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/services/projects.services.ts [hunk 7]
```diff
@@ -829,4 +829,84 @@ export default class ProjectsService {
...
+  ): Promise<Material> {
+    const material = await prisma.material.findUnique({
+      where: {
+        materialId
+      }
+    });
+
+    if (!material) throw new NotFoundException('Material', materialId);
+    if (material.dateDeleted) throw new DeletedException('Material', materialId);
+
```
Comment (RChandler234) [inline src/backend/src/services/projects.services.ts:876]: you also need all these checks from the createMaterialEndpoint, I'd recommend putting them into a helper method or smthn:
```
    if (assemblyId) {
      const assembly = await prisma.assembly.findFirst({ where: { assemblyId } });
      if (!assembly) throw new NotFoundException('Assembly', assemblyId);
    }

    const materialType = await prisma.material_Type.findFirst({
      where: { name: materialTypeName }
    });
    if (!materialType) throw new NotFoundException('Material Type', materialTypeName);

    const manufacturer = await prisma.manufacturer.findFirst({
      where: { name: manufacturerName }
    });
    if (!manufacturer) throw new NotFoundException('Manufacturer', manufacturerName);

    const unit = await prisma.unit.findFirst({
      where: { name: unitName }
    });
    if (!unit) throw new NotFoundException('Unit', unitName);
```

File: src/backend/src/services/projects.services.ts [hunk 8]
```diff
@@ -829,4 +829,84 @@ export default class ProjectsService {
...
+        manufacturerPartNumber,
+        quantity,
+        unitName,
+        price,
+        subtotal,
+        linkUrl,
+        notes,
+        wbsElementId: project.wbsElementId,
+        assemblyId: assemblyId || undefined,
+        pdmFileName: pdmFileName || undefined
```
Comment (RChandler234) [inline src/backend/src/services/projects.services.ts:906]: you don't need `: assemblyId || undefined` for that one or pdfmFileName, you can just write assemblyId and it'll do the || undefined for you

**Review by RChandler234 — CHANGES_REQUESTED** [review]

**Comment by walker-sean** [comment]
Tangentially related - could you please change material names to **not** be unique in the schema

**Review by RChandler234 — DISMISSED** [review]

**Review by RChandler234 — APPROVED** [review]
LGTM 🔥

**Pull request merged**